File: bot/commands/edit-discord-user.ts
import {UlinaCommand} from "../bot";
import * as database from "../../server/database";
import {CommandInteraction, User} from "discord.js";
import {NationAndSocials} from "../../server/database";
import {editData, editExecute, EditType, replyOk} from "../nation-edit";
import {USER_ID} from "../identifiers";
import {OkVoid} from "../../shared/result";
import {SlashCommandBuilder} from "@discordjs/builders";

async function execute(interaction: CommandInteraction, nation: NationAndSocials){
    const user: User = interaction.options.getUser(USER_ID);

    const info: database.UpdateNation = {
        nationId: nation.nationId,
        ownerDiscord: user.id
    }

    const updated = await database.updateNation(info, true);
    replyOk(interaction, updated.name);
    return OkVoid();
}

const edit: EditType = {
    name: "discord user",
    execute,
    adminOnly: true
}

module.exports = {
    data: editData(edit,new SlashCommandBuilder())
        .addUserOption(input =>
            input.setName(USER_ID)
                .setDescription("the user that owns this nation")
                .setRequired(true)
        ),
    execute: editExecute(edit),
    category: "Edit Nation"
} as UlinaCommand;

File: server/src/discord/commands/edit_discord_user.rs
use crate::database::{db, nation_change};
use crate::discord::commands::shared::{Category, CommandData, CreateCommand, Interaction};
use crate::discord::helper::Helper;
use crate::discord::ids::USER;
use crate::error::Error;
use crate::get_options;
use common::ChangeType;
use common::Id;
use serenity::client::Context;
use sqlx::query;

use super::shared::{default_data, edit_action, OptionType};

pub const DATA: CommandData = CommandData {
    admin_only: true,
    name: "edit-discord-user",
    category: Category::EditNation,
};

pub fn create(command: &mut CreateCommand) -> &mut CreateCommand {
    default_data(command, &DATA).create_option(|option| {
        option
            .kind(OptionType::User)
            .name(USER)
            .description("the user that owns this nation")
            .required(true)
    })
}

pub async fn edit_discord_user(ctx: &Context, interaction: &Interaction) -> Result<(), Error> {
    let nation = edit_action(&ctx.http, interaction, &DATA).await?;
    let user = get_options!(interaction.data.options, USER, User)?
        .id
        .0
        .to_string();

    query!(
        "UPDATE Nation SET ownerDiscord = ? WHERE nationId = ?",
        user,
        nation.nationId
    )
    .execute(db())
    .await?;

    nation_change(
        nation.id(),
        ChangeType::OwnerDiscord,
        Some(nation.ownerDiscord),
        Some(user),
        true,
    )
    .await?;

    interaction
        .message(&ctx.http, |data| data.content("successfully updated owner"))
        .await;

    Ok(())
}
